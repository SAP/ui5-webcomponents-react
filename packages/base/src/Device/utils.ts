import { getOS, supportTouch } from '@ui5/webcomponents-base/dist/Device.js';
import { supportMatchMedia, supportOrientation } from './Support';

export const getWindowSize = () => {
  return [window.innerWidth, window.innerHeight];
};

export const changeRootCSSClass = (sClassName, bRemove, sPrefix?) => {
  const oRoot = document.documentElement;
  if (oRoot.className.length === 0) {
    if (!bRemove) {
      oRoot.className = sClassName;
    }
  } else {
    const aCurrentClasses = oRoot.className.split(' ');
    let sNewClasses = '';
    aCurrentClasses.forEach((currentClass) => {
      // eslint-disable-next-line @typescript-eslint/prefer-string-starts-ends-with
      if ((sPrefix && currentClass.indexOf(sPrefix) !== 0) || (!sPrefix && currentClass !== sClassName)) {
        sNewClasses = `${sNewClasses + currentClass} `;
      }
    });
    if (!bRemove) {
      sNewClasses += sClassName;
    }
    oRoot.className = sNewClasses;
  }
};

export const getQuery = (from: number, to: number, unit = 'px') => {
  let q = 'all';
  if (from > 0) {
    q = `${q} and (min-width:${from}${unit})`;
  }
  if (to > 0) {
    q = `${q} and (max-width:${to}${unit})`;
  }
  return q;
};

export const convertToPx = (val, unit?) => {
  if (unit === 'em' || unit === 'rem') {
    const s = window.getComputedStyle ?? ((element: any) => element.currentStyle);
    const x = s(document.documentElement).fontSize;
    const f = x && x.indexOf('px') >= 0 ? parseFloat(x) : 16;
    return val * f;
  }
  return val;
};

export const matchLegacyBySize = (from, to, unit, size) => {
  from = convertToPx(from, unit);
  to = convertToPx(to, unit);

  const width = size[0];
  const a = from < 0 || from <= width;
  const b = to < 0 || width <= to;
  return a && b;
};

export const isLandscape = (bFromOrientationChange: boolean, orientation, bKeyboardOpen: boolean) => {
  if (supportTouch() && supportOrientation() && getOS().android) {
    // if on screen keyboard is open and the call of this method is from orientation change listener,
    // reverse the last value. this is because when keyboard opens on android device, the height can be less
    // than the width even in portrait mode.
    if (bKeyboardOpen && bFromOrientationChange) {
      return !orientation.landscape;
    }
    if (bKeyboardOpen) {
      // when keyboard opens, the last orientation change value will be returned.
      return orientation.landscape;
    }
  } else if (supportMatchMedia() && supportOrientation()) {
    // most desktop browsers and windows phone/tablet which not support orientationchange
    return !!window.matchMedia('(orientation: landscape)').matches;
  }
  // otherwise compare the width and height of window
  const size = getWindowSize();
  return size[0] > size[1];
};
