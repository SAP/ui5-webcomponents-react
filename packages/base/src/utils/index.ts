import { UIEvent } from 'react';

export const deprecationNotice = (component: string, message: string) => {
  if (process.env.NODE_ENV === 'development' || process.env.NODE_ENV === 'test') {
    const value = `*** ui5-webcomponents-react Deprecation Notice - ${component} ***\n`;
    // eslint-disable-next-line no-console
    if (console && console.warn) {
      // eslint-disable-next-line no-console
      console.warn(`${value}${message}`);
    }
  }
};

export const enrichEventWithDetails = <T extends Record<string, unknown>, ReturnType = CustomEvent<T>>(
  event: UIEvent,
  payload: T | null = null
) => {
  if (event.hasOwnProperty('persist')) {
    // if there is a persist method, it's an SyntheticEvent so we need to persist it
    event.persist();
  }

  const shouldCreateNewDetails =
    event.detail === null || event.detail === undefined || typeof event.detail !== 'object';
  Object.defineProperty(event, 'detail', {
    value: shouldCreateNewDetails ? {} : event.detail,
    writable: true,
    configurable: true
  });
  Object.assign(event.detail, payload);
  return event as unknown as ReturnType;
};

export { debounce } from './debounce';
