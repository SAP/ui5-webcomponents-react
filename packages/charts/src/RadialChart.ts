import { deprecationNotice } from '@ui5/webcomponents-react-base';
import { RadialChart } from './components/RadialChart/RadialChart';

deprecationNotice(
  'RadialChart',
  `importing from '@ui5/webcomponents-react-charts/dist/RadialChart' is deprecated. Please use "import { RadialChart } from '@ui5/webcomponents-react-charts';" instead.`
);
export { RadialChart };
