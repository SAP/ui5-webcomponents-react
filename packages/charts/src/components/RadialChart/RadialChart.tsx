import { CommonProps } from '@ui5/webcomponents-react';
import { enrichEventWithDetails, ThemingParameters } from '@ui5/webcomponents-react-base';
import React, { CSSProperties, forwardRef, Ref } from 'react';
import { PolarAngleAxis, RadialBar, RadialBarChart } from 'recharts';
import { useOnClickInternal } from '../../hooks/useOnClickInternal';
import { ChartContainer } from '../../internal/ChartContainer';
import { PieChartPlaceholder } from '../PieChart/Placeholder';

// todo remove this when recharts export the `CategoricalChartProps` interface themselves
interface RadialChartConfig {
  margin?: { top?: number; right?: number; bottom?: number; left?: number };
  innerRadius?: number | string;
  outerRadius?: number | string;

  [rest: string]: any;
}

export interface RadialChartProps extends Omit<CommonProps, 'onClick' | 'children' | 'onLegendClick'> {
  /**
   * The actual value which defines how much the ring is filled.
   */
  value?: number;
  /**
   * The maximum value of the ring. If `value` >= `maxValue`, the ring will be filled to 100%.
   *
   * __Default:__ `100`.
   */
  maxValue?: number;
  /**
   * The value that should be displayed in the center of the `RadialChart`.
   */
  displayValue?: number | string;
  /**
   * Font size of the `displayValue`.
   *
   * __Default values:__
   *
   * - fontSize: `1.25rem`
   * - fill: `ThemingParameters.sapTextColor`
   */
  displayValueStyle?: CSSProperties;
  /**
   * A custom color you want to apply to the ring fill. This props accepts any valid CSS color or CSS variable.
   */
  color?: CSSProperties['color'];
  /**
   * `onDataPointClick` fires when the user clicks on the filled part of the ring.
   */
  onDataPointClick?: (event: CustomEvent<{ value: number; payload: unknown; dataIndex: number }>) => void;
  /**
   * Fired when clicked anywhere in the chart.
   */
  onClick?: (
    event: CustomEvent<{
      payload: unknown;
      activePayloads: Record<string, unknown>[];
      dataIndex: number;
      value: number;
    }>
  ) => void;
  /**
   * `noAnimation` disables all chart animations when set to `true`.
   */
  noAnimation?: boolean;
  /**
   * Defines possible configurations of the internally used [RadialBarChart](https://recharts.org/en-US/api/RadialBarChart).
   *
   * __Note:__ It is possible to overwrite internally used props. Please use with caution!
   *
   * __Default values:__
   *
   * - margin: `{ top: 5, right: 5, bottom: 5, left: 5 }`
   * - innerRadius: `"90%"`
   * - outerRadius: `"100%"`
   */
  chartConfig?: RadialChartConfig;
}

const radialBarBackground = { fill: ThemingParameters.sapContent_ImagePlaceholderBackground };

const defaultDisplayValueStyles = {
  fontSize: '1.25rem',
  fill: ThemingParameters.sapTextColor,
  fontFamily: ThemingParameters.sapFontFamily
};
/**
 * Displays a ring chart highlighting a current status.
 * The status can be emphasized by using the `color` prop.
 */
const RadialChart = forwardRef((props: RadialChartProps, ref: Ref<HTMLDivElement>) => {
  const {
    maxValue,
    value,
    displayValue,
    onDataPointClick,
    onClick,
    color,
    style,
    className,
    slot,
    noAnimation,
    chartConfig,
    displayValueStyle,
    ...rest
  } = props;

  const range = [0, maxValue];
  const dataset = [{ value }];

  const onDataPointClickInternal = (payload, i, event) => {
    if (payload && onDataPointClick) {
      onDataPointClick(
        enrichEventWithDetails(event, {
          value: payload.value,
          payload: payload.payload,
          dataIndex: i
        })
      );
    }
  };

  const onClickInternal = useOnClickInternal(onClick);

  return (
    <ChartContainer
      dataset={dataset}
      ref={ref}
      Placeholder={PieChartPlaceholder}
      className={className}
      slot={slot}
      resizeDebounce={250}
      style={style}
      {...rest}
    >
      <RadialBarChart
        onClick={onClickInternal}
        innerRadius="90%"
        outerRadius="100%"
        barSize={10}
        data={dataset}
        cx="50%"
        cy="50%"
        startAngle={90}
        endAngle={-270}
        {...chartConfig}
      >
        <PolarAngleAxis type="number" domain={range} tick={false} />
        <RadialBar
          isAnimationActive={noAnimation === false}
          background={radialBarBackground}
          dataKey="value"
          fill={color ?? ThemingParameters.sapChart_OrderedColor_1}
          onClick={onDataPointClickInternal}
        />
        {displayValue && (
          <text
            x="50%"
            y="50%"
            textAnchor="middle"
            dominantBaseline="middle"
            className="progress-label"
            style={{ ...defaultDisplayValueStyles, ...displayValueStyle }}
          >
            {displayValue}
          </text>
        )}
      </RadialBarChart>
    </ChartContainer>
  );
});

RadialChart.defaultProps = {
  maxValue: 100,
  noAnimation: false,
  displayValueStyle: defaultDisplayValueStyles
};

RadialChart.displayName = 'RadialChart';

export { RadialChart };
