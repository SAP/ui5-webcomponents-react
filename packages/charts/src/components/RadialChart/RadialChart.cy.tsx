import { RadialChart } from './RadialChart.js';
import { cypressPassThroughTestsFactory } from '@/cypress/support/utils';

describe('RadialChart', () => {
  it('Basic', () => {
    cy.mount(<RadialChart value={67} displayValue={'67%'} />);
    cy.get('.recharts-responsive-container').should('be.visible');
    cy.get('.recharts-area').should('have.length', 1);
    cy.get('.recharts-radial-bar-sectors').should('have.length', 1);
    cy.findByText('67%').should('be.visible');
  });

  it('click handlers', () => {
    const onClick = cy.spy().as('onClick');
    cy.mount(<RadialChart value={67} displayValue={'67%'} onDataPointClick={onClick} />);

    cy.get('.recharts-radial-bar-sector').click();
    cy.get('@onClick')
      .should('have.been.calledOnce')
      .and(
        'have.been.calledWith',
        Cypress.sinon.match({
          detail: Cypress.sinon.match({
            payload: { value: 67 }
          })
        })
      );
  });

  cypressPassThroughTestsFactory(RadialChart, { value: 67, displayValue: '67%' });
});
