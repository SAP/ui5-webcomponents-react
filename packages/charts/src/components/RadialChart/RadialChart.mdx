import { ControlsWithNote, DocsHeader, Footer } from '@sb/components';
import { Canvas, Meta } from '@storybook/addon-docs';
import * as ComponentStories from './RadialChart.stories';

<Meta of={ComponentStories} />

<DocsHeader />

## Example

<Canvas of={ComponentStories.Default} />

## Properties

<ControlsWithNote of={ComponentStories.Default} />

<br />
<br />

## More Examples

### With Custom Color

<Canvas of={ComponentStories.WithCustomColor} />

### Micro RadialCharts

<Canvas of={ComponentStories.MicroRadialCharts} />

<Footer />
