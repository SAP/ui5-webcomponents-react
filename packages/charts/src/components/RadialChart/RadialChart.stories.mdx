import { ArgsTable, Canvas, Meta, Story } from '@storybook/addon-docs';
import { DocsHeader } from '@docs/DocsHeader';
import { Footer } from '@docs/Footer';
import { RadialChart } from '@ui5/webcomponents-react-charts/dist/RadialChart';

<Meta
  title="Charts /  RadialChart"
  component={RadialChart}
  args={{
    displayValue: '50%',
    value: 50,
    maxValue: 100
  }}
/>

<DocsHeader />

## Example

<Canvas>
  <Story name="Default">{(props) => <RadialChart {...props} />}</Story>
</Canvas>

## Properties

<ArgsTable story="Default" />

<br />
<br />

## More Examples

### With Custom Color

<Canvas>
  <Story name="With Custom Color" args={{ color: '#f0ab00' }}>
    {(props) => <RadialChart {...props} />}
  </Story>
</Canvas>

<Footer />
