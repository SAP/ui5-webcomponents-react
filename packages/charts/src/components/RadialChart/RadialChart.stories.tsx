import type { Meta, StoryObj } from '@storybook/react';
import { FlexBox, FlexBoxAlignItems, FlexBoxJustifyContent, Text } from '@ui5/webcomponents-react';
import { ThemingParameters } from '@ui5/webcomponents-react-base';
import React from 'react';
import { RadialChart } from './RadialChart.js';

const meta = {
  title: 'RadialChart',
  component: RadialChart,
  args: { displayValue: '50%', value: 50, maxValue: 100 }
} satisfies Meta<typeof RadialChart>;

export default meta;
type Story = StoryObj<typeof meta>;

export const Default: Story = {};

export const WithCustomColor: Story = {
  args: {
    color: '#f0ab00'
  }
};

export const MicroRadialCharts: Story = {
  render: () => {
    return (
      <FlexBox justifyContent={FlexBoxJustifyContent.SpaceBetween} alignItems={FlexBoxAlignItems.Center}>
        <RadialChart
          value={50}
          style={{ height: '50px', width: '50px' }}
          chartConfig={{ innerRadius: '75%', margin: { top: 0, right: 0, bottom: 0, left: 0 } }}
          displayValue={'50%'}
          displayValueStyle={{
            fontSize: ThemingParameters.sapFontSmallSize,
            fill: ThemingParameters.sapChart_OrderedColor_1
          }}
        />
        <FlexBox alignItems={FlexBoxAlignItems.Center}>
          <RadialChart
            value={50}
            style={{ width: '25px', height: '25px' }}
            chartConfig={{ innerRadius: '70%', margin: { top: 0, right: 0, bottom: 0, left: 0 } }}
          />
          <Text>50%</Text>
        </FlexBox>
        <FlexBox alignItems={FlexBoxAlignItems.Center}>
          <RadialChart
            value={50}
            style={{ height: '16px', width: '16px' }}
            chartConfig={{ innerRadius: '75%', margin: { top: 0, right: 0, bottom: 0, left: 0 } }}
          />
          <Text style={{ fontSize: ThemingParameters.sapFontSmallSize }}>50%</Text>
        </FlexBox>
      </FlexBox>
    );
  }
};
