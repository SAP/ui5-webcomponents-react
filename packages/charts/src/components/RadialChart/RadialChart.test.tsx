import { createChartRenderTest, createOnClickChartTest } from '@shared/tests/chartUtils';
import { createCustomPropsTest } from '@shared/tests/utils';
import { RadialChart } from './RadialChart';

const value = 67;

const displayValue = '67%';

describe('RadialChart', () => {
  createChartRenderTest(RadialChart, { value, displayValue });

  createOnClickChartTest(RadialChart, { value, displayValue, noLegend: true });

  createCustomPropsTest(RadialChart, { value, displayValue });
});
