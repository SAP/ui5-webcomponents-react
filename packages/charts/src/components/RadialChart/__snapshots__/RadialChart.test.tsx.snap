// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RadialChart Check onClick events 1`] = `
<DocumentFragment>
  <div
    style="font-family: var(--sapFontFamily); width: 100%; height: 400px; position: relative;"
  >
    <div
      class="recharts-responsive-container"
      style="width: 400px; height: 400px;"
    >
      <div
        class="recharts-wrapper"
        style="position: relative; cursor: default; width: 400px; height: 400px;"
      >
        <svg
          class="recharts-surface"
          cx="50%"
          cy="50%"
          height="400"
          version="1.1"
          viewBox="0 0 400 400"
          width="400"
        >
          <defs>
            <clippath
              id="recharts3-clip"
            >
              <rect
                height="340"
                width="340"
                x="30"
                y="30"
              />
            </clippath>
          </defs>
          <g
            class="recharts-layer recharts-polar-angle-axis"
          >
            <path
              class="recharts-polygon angleAxis"
              cx="200"
              cy="200"
              d="M200,30L299.9234928897204,62.46711095625892L361.67960777017606,147.46711095625895L361.67960777017606,252.53288904374108L299.9234928897204,337.5328890437411L200,370L100.07650711027958,337.5328890437411L38.32039222982391,252.53288904374108L38.32039222982388,147.46711095625898L100.07650711027955,62.46711095625895L199.99999999999997,30L200,30Z"
              fill="none"
              orientation="outer"
              radius="170"
              type="number"
            />
            <g
              class="recharts-layer recharts-polar-angle-axis-ticks"
            >
              <g
                class="recharts-layer recharts-polar-angle-axis-tick"
              >
                <line
                  class="angleAxis"
                  cx="200"
                  cy="200"
                  fill="none"
                  orientation="outer"
                  radius="170"
                  type="number"
                  x1="200"
                  x2="200"
                  y1="30"
                  y2="22"
                />
              </g>
              <g
                class="recharts-layer recharts-polar-angle-axis-tick"
              >
                <line
                  class="angleAxis"
                  cx="200"
                  cy="200"
                  fill="none"
                  orientation="outer"
                  radius="170"
                  type="number"
                  x1="299.9234928897204"
                  x2="304.62577490806024"
                  y1="62.46711095625892"
                  y2="55.99497500125935"
                />
              </g>
              <g
                class="recharts-layer recharts-polar-angle-axis-tick"
              >
                <line
                  class="angleAxis"
                  cx="200"
                  cy="200"
                  fill="none"
                  orientation="outer"
                  radius="170"
                  type="number"
                  x1="361.67960777017606"
                  x2="369.2880599005373"
                  y1="147.46711095625895"
                  y2="144.99497500125938"
                />
              </g>
              <g
                class="recharts-layer recharts-polar-angle-axis-tick"
              >
                <line
                  class="angleAxis"
                  cx="200"
                  cy="200"
                  fill="none"
                  orientation="outer"
                  radius="170"
                  type="number"
                  x1="361.67960777017606"
                  x2="369.2880599005373"
                  y1="252.53288904374108"
                  y2="255.00502499874065"
                />
              </g>
              <g
                class="recharts-layer recharts-polar-angle-axis-tick"
              >
                <line
                  class="angleAxis"
                  cx="200"
                  cy="200"
                  fill="none"
                  orientation="outer"
                  radius="170"
                  type="number"
                  x1="299.9234928897204"
                  x2="304.62577490806024"
                  y1="337.5328890437411"
                  y2="344.0050249987406"
                />
              </g>
              <g
                class="recharts-layer recharts-polar-angle-axis-tick"
              >
                <line
                  class="angleAxis"
                  cx="200"
                  cy="200"
                  fill="none"
                  orientation="outer"
                  radius="170"
                  type="number"
                  x1="200"
                  x2="200"
                  y1="370"
                  y2="378"
                />
              </g>
              <g
                class="recharts-layer recharts-polar-angle-axis-tick"
              >
                <line
                  class="angleAxis"
                  cx="200"
                  cy="200"
                  fill="none"
                  orientation="outer"
                  radius="170"
                  type="number"
                  x1="100.07650711027958"
                  x2="95.3742250919398"
                  y1="337.5328890437411"
                  y2="344.0050249987406"
                />
              </g>
              <g
                class="recharts-layer recharts-polar-angle-axis-tick"
              >
                <line
                  class="angleAxis"
                  cx="200"
                  cy="200"
                  fill="none"
                  orientation="outer"
                  radius="170"
                  type="number"
                  x1="38.32039222982391"
                  x2="30.71194009946268"
                  y1="252.53288904374108"
                  y2="255.00502499874065"
                />
              </g>
              <g
                class="recharts-layer recharts-polar-angle-axis-tick"
              >
                <line
                  class="angleAxis"
                  cx="200"
                  cy="200"
                  fill="none"
                  orientation="outer"
                  radius="170"
                  type="number"
                  x1="38.32039222982388"
                  x2="30.71194009946265"
                  y1="147.46711095625898"
                  y2="144.99497500125938"
                />
              </g>
              <g
                class="recharts-layer recharts-polar-angle-axis-tick"
              >
                <line
                  class="angleAxis"
                  cx="200"
                  cy="200"
                  fill="none"
                  orientation="outer"
                  radius="170"
                  type="number"
                  x1="100.07650711027955"
                  x2="95.37422509193976"
                  y1="62.46711095625895"
                  y2="55.99497500125938"
                />
              </g>
              <g
                class="recharts-layer recharts-polar-angle-axis-tick"
              >
                <line
                  class="angleAxis"
                  cx="200"
                  cy="200"
                  fill="none"
                  orientation="outer"
                  radius="170"
                  type="number"
                  x1="199.99999999999997"
                  x2="199.99999999999997"
                  y1="30"
                  y2="22"
                />
              </g>
            </g>
          </g>
          <g
            class="recharts-layer recharts-area"
          >
            <g
              class="recharts-layer recharts-radial-bar-background"
            >
              <path
                class="recharts-sector recharts-radial-bar-background-sector"
                cx="200"
                cy="200"
                d="M 200,42.5
    A 157.5,157.5,0,
    1,1,
    199.99725110642825,42.50000002398863
  L 199.99742563935345,52.50000002246554
            A 147.5,147.5,0,
            1,0,
            200,52.5 Z"
                fill="var(--sapContent_ImagePlaceholderBackground)"
              />
            </g>
            <g
              class="recharts-layer recharts-radial-bar-sectors"
            >
              <g
                class="recharts-layer"
              />
            </g>
          </g>
          <text
            class="progress-label"
            dominant-baseline="middle"
            style="fill: var(--sapTextColor);"
            text-anchor="middle"
            x="50%"
            y="50%"
          >
            67%
          </text>
        </svg>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`RadialChart Render chart with data 1`] = `
<DocumentFragment>
  <div
    style="font-family: var(--sapFontFamily); width: 100%; height: 400px; position: relative;"
  >
    <div
      class="recharts-responsive-container"
      style="width: 400px; height: 400px;"
    >
      <div
        class="recharts-wrapper"
        style="position: relative; cursor: default; width: 400px; height: 400px;"
      >
        <svg
          class="recharts-surface"
          cx="50%"
          cy="50%"
          height="400"
          version="1.1"
          viewBox="0 0 400 400"
          width="400"
        >
          <defs>
            <clippath
              id="recharts1-clip"
            >
              <rect
                height="340"
                width="340"
                x="30"
                y="30"
              />
            </clippath>
          </defs>
          <g
            class="recharts-layer recharts-polar-angle-axis"
          >
            <path
              class="recharts-polygon angleAxis"
              cx="200"
              cy="200"
              d="M200,30L299.9234928897204,62.46711095625892L361.67960777017606,147.46711095625895L361.67960777017606,252.53288904374108L299.9234928897204,337.5328890437411L200,370L100.07650711027958,337.5328890437411L38.32039222982391,252.53288904374108L38.32039222982388,147.46711095625898L100.07650711027955,62.46711095625895L199.99999999999997,30L200,30Z"
              fill="none"
              orientation="outer"
              radius="170"
              type="number"
            />
            <g
              class="recharts-layer recharts-polar-angle-axis-ticks"
            >
              <g
                class="recharts-layer recharts-polar-angle-axis-tick"
              >
                <line
                  class="angleAxis"
                  cx="200"
                  cy="200"
                  fill="none"
                  orientation="outer"
                  radius="170"
                  type="number"
                  x1="200"
                  x2="200"
                  y1="30"
                  y2="22"
                />
              </g>
              <g
                class="recharts-layer recharts-polar-angle-axis-tick"
              >
                <line
                  class="angleAxis"
                  cx="200"
                  cy="200"
                  fill="none"
                  orientation="outer"
                  radius="170"
                  type="number"
                  x1="299.9234928897204"
                  x2="304.62577490806024"
                  y1="62.46711095625892"
                  y2="55.99497500125935"
                />
              </g>
              <g
                class="recharts-layer recharts-polar-angle-axis-tick"
              >
                <line
                  class="angleAxis"
                  cx="200"
                  cy="200"
                  fill="none"
                  orientation="outer"
                  radius="170"
                  type="number"
                  x1="361.67960777017606"
                  x2="369.2880599005373"
                  y1="147.46711095625895"
                  y2="144.99497500125938"
                />
              </g>
              <g
                class="recharts-layer recharts-polar-angle-axis-tick"
              >
                <line
                  class="angleAxis"
                  cx="200"
                  cy="200"
                  fill="none"
                  orientation="outer"
                  radius="170"
                  type="number"
                  x1="361.67960777017606"
                  x2="369.2880599005373"
                  y1="252.53288904374108"
                  y2="255.00502499874065"
                />
              </g>
              <g
                class="recharts-layer recharts-polar-angle-axis-tick"
              >
                <line
                  class="angleAxis"
                  cx="200"
                  cy="200"
                  fill="none"
                  orientation="outer"
                  radius="170"
                  type="number"
                  x1="299.9234928897204"
                  x2="304.62577490806024"
                  y1="337.5328890437411"
                  y2="344.0050249987406"
                />
              </g>
              <g
                class="recharts-layer recharts-polar-angle-axis-tick"
              >
                <line
                  class="angleAxis"
                  cx="200"
                  cy="200"
                  fill="none"
                  orientation="outer"
                  radius="170"
                  type="number"
                  x1="200"
                  x2="200"
                  y1="370"
                  y2="378"
                />
              </g>
              <g
                class="recharts-layer recharts-polar-angle-axis-tick"
              >
                <line
                  class="angleAxis"
                  cx="200"
                  cy="200"
                  fill="none"
                  orientation="outer"
                  radius="170"
                  type="number"
                  x1="100.07650711027958"
                  x2="95.3742250919398"
                  y1="337.5328890437411"
                  y2="344.0050249987406"
                />
              </g>
              <g
                class="recharts-layer recharts-polar-angle-axis-tick"
              >
                <line
                  class="angleAxis"
                  cx="200"
                  cy="200"
                  fill="none"
                  orientation="outer"
                  radius="170"
                  type="number"
                  x1="38.32039222982391"
                  x2="30.71194009946268"
                  y1="252.53288904374108"
                  y2="255.00502499874065"
                />
              </g>
              <g
                class="recharts-layer recharts-polar-angle-axis-tick"
              >
                <line
                  class="angleAxis"
                  cx="200"
                  cy="200"
                  fill="none"
                  orientation="outer"
                  radius="170"
                  type="number"
                  x1="38.32039222982388"
                  x2="30.71194009946265"
                  y1="147.46711095625898"
                  y2="144.99497500125938"
                />
              </g>
              <g
                class="recharts-layer recharts-polar-angle-axis-tick"
              >
                <line
                  class="angleAxis"
                  cx="200"
                  cy="200"
                  fill="none"
                  orientation="outer"
                  radius="170"
                  type="number"
                  x1="100.07650711027955"
                  x2="95.37422509193976"
                  y1="62.46711095625895"
                  y2="55.99497500125938"
                />
              </g>
              <g
                class="recharts-layer recharts-polar-angle-axis-tick"
              >
                <line
                  class="angleAxis"
                  cx="200"
                  cy="200"
                  fill="none"
                  orientation="outer"
                  radius="170"
                  type="number"
                  x1="199.99999999999997"
                  x2="199.99999999999997"
                  y1="30"
                  y2="22"
                />
              </g>
            </g>
          </g>
          <g
            class="recharts-layer recharts-area"
          >
            <g
              class="recharts-layer recharts-radial-bar-background"
            >
              <path
                class="recharts-sector recharts-radial-bar-background-sector"
                cx="200"
                cy="200"
                d="M 200,42.5
    A 157.5,157.5,0,
    1,1,
    199.99725110642825,42.50000002398863
  L 199.99742563935345,52.50000002246554
            A 147.5,147.5,0,
            1,0,
            200,52.5 Z"
                fill="var(--sapContent_ImagePlaceholderBackground)"
              />
            </g>
            <g
              class="recharts-layer recharts-radial-bar-sectors"
            >
              <g
                class="recharts-layer"
              />
            </g>
          </g>
          <text
            class="progress-label"
            dominant-baseline="middle"
            style="fill: var(--sapTextColor);"
            text-anchor="middle"
            x="50%"
            y="50%"
          >
            67%
          </text>
        </svg>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
